### PR TITLE
gh-102511: Add C implementation of `os.path.splitroot()`

### PR DESCRIPTION
**Pull request by nineteendo**
#### ntpath.py

```powershell
PS C:\Users\wanne\cpython> main\python -m timeit -s "import os" "os.path.splitroot('//server/share/spam/eggs')"; speedup-os.path.splitroot\python -m timeit -s "import os" "os.path.splitroot('//server/share/spam/eggs')"
200000 loops, best of 5: 1.14 usec per loop # before
500000 loops, best of 5: 433 nsec per loop # after
# -> 2.63x faster
```

* Issue: gh-102511

### CODE REVIEW

File: Modules/posixmodule.c
Comment (erlend-aasland) [inline Modules/posixmodule.c]: You should handle errors for all the `PyUnicode_` API calls. If `NULL` is returned, clean up and bail.

File: Python/fileutils.c
Comment (zooba) [inline Python/fileutils.c:2301]: At the very least we should `assert(drvsize)` and `assert(rootsize)`.

File: Python/fileutils.c
Comment (zooba) [inline Python/fileutils.c]: Any harm in `const wchar_t*` here?

Comment (zooba) [inline Python/fileutils.c]: I wonder if there's any value in having a "type" argument here to choose between POSIX and Windows handling at runtime (with either `0` or `-1` indicating "native")? That way we can have x-plat fallback behaviour use the same implementation, and we have less code to keep in sync.

File: Lib/importlib/_bootstrap_external.py
Comment (zooba) [inline Lib/importlib/_bootstrap_external.py]: Previously this could handle `Path` objects (and people tried it by adding them to `sys.path`, which has a ~50/50 chance of ever working). Now it will fail early enough to reduce that to a 0% chance of working. IIRC, we don't have `os.fsdecode` here, but using `path_t` in the native function works correctly.

I'm not totally against that, but it's worth understanding the situations this will impact and making sure they're correctly documented.

File: Misc/NEWS.d/next/Core and Builtins/2024-04-19-08-50-48.gh-issue-102511.qDEB66.rst
Comment (zooba) [inline Misc/NEWS.d/next/Core and Builtins/2024-04-19-08-50-48.gh-issue-102511.qDEB66.rst]:
```suggestion
Speed up :func:`os.path.splitroot` with a native implementation.
```

File: Modules/posixmodule.c
Comment (zooba) [inline Modules/posixmodule.c:5473]: (See earlier comment - this function is called from places where `os.fsdecode` is not available, so using `path_t` may be the only way to preserve existing behaviours.)

**Review by zooba — COMMENTED** [review]
Great looking change! As Erlend pointed out, a couple of robustness things to check, and there's a potential behaviour change that's worth validating and describing in `Doc\whatsnew\3.13.rst` and possibly in other parts of the documentation.

File: Lib/importlib/_bootstrap_external.py
Comment (nineteendo) [reply Lib/importlib/_bootstrap_external.py]: I've restored the old implementation and renamed it to `_path_splitanchor`. Let's leave removing it for another time.

File: Python/fileutils.c
Comment (nineteendo) [reply Python/fileutils.c]: That's beyond the scope of this pull request. Feel free to make an issue for fixing this in a followup though.

File: Lib/importlib/_bootstrap_external.py
Comment (zooba) [reply Lib/importlib/_bootstrap_external.py]: It's literally just the use of `path_t` instead of `unicode` in the new definition. I know that adds an extra few checks in C code to covert back to bytes, but that saves the same check currently in Python code and it's really not that difficult (you'll see it in most other functions already).

But if we're going to leave the old one there, leave it with its original name and make the new function a new function.

File: Modules/posixmodule.c
Comment (nineteendo) [inline Modules/posixmodule.c]: What do we do with the docstring here? There's no longer a Python wrapper around this function.

File: Lib/importlib/_bootstrap_external.py
Comment (nineteendo) [reply Lib/importlib/_bootstrap_external.py]: I think I figured it out, but now I'm getting a segmentation fault on Unix.

**Comment by nineteendo** [comment]
Ehm, why am I getting a segmentation fault?

**Comment by nineteendo** [comment]
Oh, come on, why must argclinic raise an error here? That isn't useful:

```none
ValueError: _path_splitroot_ex: embedded null character in p
```

**Comment by zooba** [comment]
> _path_splitroot_ex: embedded null character in p

Oh right, I forgot about this behaviour of argument clinic 😞 Might be a good reason to go to `unicode` after all.

Alternatively, you could handle the `ValueError` in the Python wrapper, split at the null, and then append it back onto the last element? Strictly speaking we're in the "invalid path --> invalid output" place again, as we don't actually support embedded nulls in paths at all, but there's no reason we need to go ahead and raise the error this early.

File: Modules/posixmodule.c
Comment (zooba) [reply Modules/posixmodule.c]: Worth adding a one liner for what it does, or at least the return value. It'll still be accessible from Python, and a short docstring is not the same as documented.

**Comment by nineteendo** [comment]
I added an `allow_embedded_null` argument to `path_convertor`. No need for a Python wrapper. We can add it to the struct later.

**Comment by nineteendo** [comment]
OK, I'll need a python wrapper anyway, as surrogates aren't handled correctly.

**Comment by nineteendo** [comment]
Do we still need to document a behaviour change now that bootstrap isn't changed?

**Comment by nineteendo** [comment]
@eryksun, sorry for the ping, but could you compare this with your implementation?

File: Modules/posixmodule.c
Comment (zooba) [inline Modules/posixmodule.c]: Can we `const` the earlier `buffer` definition to avoid the warnings here?

Comment (zooba) [reply Modules/posixmodule.c]: Ah, GitHub landed me on a partial commit, so I didn't see the rest of the change.

I'd rather define the variable as `const` if we can than cast away the `const`. If it needs to be non-const for non-Windows (though I don't see why), declare it inside the `#ifdef` (we no longer require all declarations at the top of the function, so it can go on the first assignment).

**Comment by zooba** [comment]
> Do we still need to document a behaviour change now that bootstrap isn't changed?

Nope

File: Modules/posixmodule.c
Comment (nineteendo) [reply Modules/posixmodule.c]: If fixed it by using a different declaration for Windows and non-Windows.

**Comment by nineteendo** [comment]
I switched back to handling bytes in the Python implementation. It's a bit slower, but less ugly.

**Comment by nineteendo** [comment]
@barneygale, since this is your issue, do you have any suggestions? Or is your C not good enough?

**Review by zooba — APPROVED** [review]
This LGTM, so if @barneygale is happy, he can merge.

File: Modules/posixmodule.c
Comment (erlend-aasland) [inline Modules/posixmodule.c]: `PyMem_Free` [is a no-op](https://docs.python.org/3/c-api/memory.html#c.PyMem_Free) if called with a `NULL` pointer; `Py_XDECREF` will do the `if (...) Py_DECREF` dance for you.

```suggestion
    PyMem_Free(buffer);
    Py_XDECREF(drv);
    Py_DECREF(root);
    Py_DECREF(tail);
```

File: Modules/posixmodule.c
Comment (erlend-aasland) [inline Modules/posixmodule.c]: This C style is not commonly used in our code base; consider spelling it out:

```suggestion
    drv = PyUnicode_FromWideChar(buffer, drvsize);
    if (drv == NULL) {
        goto exit;
    }
    root = PyUnicode_FromWideChar(&buffer[drvsize], rootsize);
    if (root == NULL) {
        goto exit;
    }
    tail = PyUnicode_FromWideChar(&buffer[drvsize + rootsize],
                                  len - drvsize - rootsize);
    if (tail == NULL) {
        goto exit;
    }
```

**Comment by barneygale** [comment]
Sorry if this is a silly question, but why not use [`PathCchSkipRoot`](https://learn.microsoft.com/en-us/windows/win32/api/pathcch/nf-pathcch-pathcchskiproot) in the implementation?

**Comment by zooba** [comment]
> why not use [`PathCchSkipRoot`](https://learn.microsoft.com/en-us/windows/win32/api/pathcch/nf-pathcch-pathcchskiproot) in the implementation

It behaves inconsistently in some cases (mostly invalid paths, though I believe it also doesn't handle `\\?\UNC\` the way it ought).

**Comment by nineteendo** [comment]
It doesn't even return the drive index...

File: Python/fileutils.c
Comment (eryksun) [inline Python/fileutils.c]: PEP 7 still limits lines to 79 characters.

```suggestion
            if (path[2] == L'?' && IS_SEP(&path[3]) &&
                (path[4] == L'U' || path[4] == L'u') &&
                (path[5] == L'N' || path[5] == L'n') &&
                (path[6] == L'C' || path[6] == L'c') &&
                IS_SEP(&path[7]))
```

File: Modules/posixmodule.c
Comment (nineteendo) [reply Modules/posixmodule.c]: How uncommon is 525 times? Regex: `if \(.* = `. Or do you mean the multiple conditions?
And I think spelling it out makes it less readable, but you have the final say.

Comment (erlend-aasland) [reply Modules/posixmodule.c]: @nineteendo, I'd appreciate if you could apply this suggestion as well. We try to keep our C code as readable as possible.

Comment (erlend-aasland) [reply Modules/posixmodule.c]: > How uncommon is 525 times?

The correct number is much lower[^1], and it really does not matter since it is still the minority.

[^1]: for example, mimalloc and libmpdec are vendored, Python-ast.c is autogenerated.

File: Python/fileutils.c
Comment (erlend-aasland) [reply Python/fileutils.c]: Please apply this suggestion as well, @nineteendo.

Comment (nineteendo) [reply Python/fileutils.c]: There's no rush, I apply all suggestions. I simply wanted to check this in my IDE.

File: Modules/posixmodule.c
Comment (nineteendo) [reply Modules/posixmodule.c]: I see, then I'll stick to the guidelines.

**Review by barneygale — APPROVED** [review]
My C isn't good enough for me to confidently check for out-of-bounds accesses and whatnot. But the algorithm looks good to me and matches the Python version.

I hereby pass the torch to @erlend-aasland to merge when he wishes!

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
I'll land this tomorrow (CET) to give @eryksun a chance to give his thumbs up.

**Comment by eryksun** [comment]
> @eryksun, sorry for the ping, but could you compare this with your implementation?

My implementation returned a pointer to the `rest` of the path, plus an optional out parameter pointer to the `root`, and `_path_splitroot_ex()` used pointer arithmetic. You took a more direct approach, with required out parameters for the drive size and root size. I think the performance should be about the same from what I see in your implementation, but I was seeing a 4x improvement on Windows, and you're only seeing 2.78x improvement. I'll check the performance of your PR versus the current main.

File: Modules/posixmodule.c
Comment (eryksun) [inline Modules/posixmodule.c:5478]: Maybe rename the old function `_path_splitroot_win` since it's based on WinAPI `PathCchSkipRoot()` and is only used on Windows in "Lib/importlib/\_bootstrap\_external.py". Then this new function can simply be named `_path_splitroot`, and we won't have to needlessly use an "ex" name.

Comment (nineteendo) [reply Modules/posixmodule.c:5478]: I initially wanted to rename it to `_path_splitanchor`, but @zooba didn't agree:

> But if we're going to leave the old one there, leave it with its original name and make the new function a new function.

I also don't like the "ex" name.

**Comment by nineteendo** [comment]
It might just be my laptop or because I'm comparing with 3.12. (there doesn't seem to be a way to have 2 builds at the same time)
Otherwise we need to check what's different from your implementation. (if you haven't lost it, of course)

File: Modules/posixmodule.c
Comment (nineteendo) [reply Modules/posixmodule.c:5478]: OK, with 3 people on board an 1 against I'll rename the functions.

Comment (eryksun) [reply Modules/posixmodule.c:5478]: Steve's opinion is a very big deal to me, considering how much he's done for Python on Windows for several years now. If he prefers the "ex" function, I'm not going to argue over something that trivial.

Comment (nineteendo) [reply Modules/posixmodule.c:5478]: OK, just in time. We keep the "ex" prefix then.

**Comment by nineteendo** [comment]
OK, how's the build going?

**Comment by eryksun** [comment]
> It might just be my laptop or because I'm comparing with 3.12. (there doesn't seem to be a way to have 2 builds at the same time)

You can use worktrees. Anyway, I still have my implementation, but there's no need to delve further into the comparison because the reason is obvious. I was testing the builtin function directly, not a wrapper function that called `os.fspath()` before calling the builtin. My plan was to switch to using `path_t` internally instead of a wrapper, but I never implemented that. Here's the performance I get for your PR with and without the wrapper.

```text
> .\python_work -m timeit -s "import os" "os.path.splitroot('//server/share/spam/eggs')"
1000000 loops, best of 5: 270 nsec per loop

> .\python_work -m timeit -s "import nt" "nt._path_splitroot_ex('//server/share/spam/eggs')"
2000000 loops, best of 5: 153 nsec per loop
```

It's 1.76 times faster without the wrapper.

**Comment by nineteendo** [comment]
> You can use worktrees.

Could you explain how I can use them?

> I was testing the builtin function directly, not a wrapper function

Yeah, that matches the performance when I used `path_t` before. But I switched back to using a Python wrapper because NULL bytes and surrogates weren't handled properly and the code was quite ugly with having to convert the input path to bytes on Unix.

**Comment by eryksun** [comment]
Here's a brief overview of worktrees in the developer's guide:

https://devguide.python.org/getting-started/git-boot-camp/#git-worktree

> I tried to work around the first issue by modifying `path_converter`, but I couldn't get surrogates to work, so I abandoned that idea and went for a simpler implementation.

In another issue, the implementation of `path_t` could be generalized to support fields to configure the converter to use `wide` regardless of platform, to allow null characters, to allow arbitrary length paths (e.g. no 32767 length limit on Windows), and a new field such as `bytes_input` to determine whether a path result has to be converted back to `bytes`. The option to always use a wide-character path is a generalization of the current behavior on Windows. Argument Clinic would be extended to support the new options. The implementations of `_path_splitroot_ex()`, `_path_normpath()`, and `_path_abspath()` (if adopted) would benefit, and also the `_path_is*()` helpers on Windows.

File: Modules/posixmodule.c
Comment (eryksun) [inline Modules/posixmodule.c]: These values may be `NULL` here.

```suggestion
    Py_XDECREF(root);
    Py_XDECREF(tail);
```

Comment (nineteendo) [reply Modules/posixmodule.c]: I forgot to change these after erlends suggestion.

**Comment by nineteendo** [comment]
Let's leave `path_t` for a followup pull request, such that it can be implemented without a Python wrapper.

**Review by eryksun — APPROVED** [review]
LGTM. Thanks, nineteendo.

**Comment by zooba** [comment]
That's quorum, thanks @nineteendo for bearing with us, and congrats on landing the patch!

**Comment by nineteendo** [comment]
Hmm, it didn't fail in the previous commit: https://buildbot.python.org/all/#/builders/1380/builds/86
Is this random?

**Comment by zooba** [comment]
This seems unrelated, and if it's only iOS, then we can ignore it. That buildbot has only just started running, and will be flushing out random issues for a little while.

**Comment by nineteendo** [comment]
Eryk, can we backport this to 3.12 to fix the bug with `ntpath.normpath()`?

**Comment by gpshead** [comment]
Please add a unittest coverage of both the Python and C versions so that we ensure their behavior is the same.

**Comment by eryksun** [comment]
@gpshead, they both get tested as long as tests run on at least one POSIX system and one Windows system. The Python fallback implementation is only defined if the C version can't be imported. Thus the fallback for `ntpath.splitroot()` is only defined on POSIX, and the fallback for `posixpath.splitroot()` is only defined on Windows.

**Review by sobolevn — COMMENTED** [review]
This PR also changes first parameter name. It is now:
- `p` for python version
- `path` for C version

**Comment by hugovk** [comment]
Is this intentional? Does it need addressing before 3.13.0 in two weeks?

**Comment by sobolevn** [comment]
I would prefer to address this, because:
- It can break a user's code that used: `os.path.splitroot(p="...")`
- It changes the `typeshed` typings

**Pull request merged**
